### PR TITLE
added tests and functionality for retire product. updated view pages …

### PR DESCRIPTION
**Pull request by aribray**
…to only show products that are not retired

# bEtsy
Congratulations! You're submitting your assignment! These comprehension questions should be answered by all members of your team, not by a single teammate.

## Comprehension Questions
Question | Answer
:------------- | :-------------
Each team member: what is one thing you were primarily responsible for that you're proud of? | 
Each team member: what is one thing you were primarily responsible for that you would like targeted feedback on? | 
How did your team break up the work to be done? | 
How did your team utilize git to collaborate? | 
What did your group do to try to keep your code DRY while many people collaborated on it? | 
What was a technical challenge that you faced as a group? | 
What was a team/personal challenge that you faced as a group? | 
What was your application's ERD? (include a link) | 
What is your Trello URL? | 
What is the Heroku URL of your deployed application? |

### CODE REVIEW

File: app/controllers/products_controller.rb [hunk 1]
```diff
@@ -1,6 +1,8 @@
+# frozen_string_literal: true
+
 class ProductsController < ApplicationController
-  before_action :find_individual_product, only: [:show, :edit, :update, :retire]
-  before_action :require_login, only: [:new, :create, :edit, :update, :retire]
+  before_action :find_individual_product, only: %i[show edit update retire]
+  before_action :require_login, only: %i[new create edit update retire]
```
Comment (jessicacaley) [inline app/controllers/products_controller.rb:5]: Ooh, cool refactor.

File: app/views/categories/show.html.erb [hunk 2]
```diff
@@ -1,23 +1,25 @@
...
-      <%= number_to_currency(product.price)%>
-    </li>
-    <li>
-      <%= product.quantity%>
-    </li>
-  </ul>
+  <% if product.retired == false%>
+    <ul>
+      <li>
+        <%=image_tag(product.photo_url)%>
```
Comment (jessicacaley) [inline app/views/categories/show.html.erb:7]: Probably should add an alt (I haven't been doing that either but we should get in the habit for accessibility)

File: app/views/products/index.html.erb [hunk 3]
```diff
@@ -2,14 +2,16 @@
...
-            <li><%= link_to product.user_id, user_path(user.id) %></li> %>
-            <% end %>
-            <li><%= link_to product.description, product_path(product.id) %></li>
-        </ul>
+        <% if product.retired == false %>
+            <img src= <%= product.photo_url %> alt="image of #{product.name}">
+            <ul>
+                <li><h3><%= link_to product.name, product_path(product.id) %></h3></li>
+                <% if false %>
+                currently commented out.  need help with pathway to user/product_id
```
Comment (jessicacaley) [inline app/views/products/index.html.erb]: I think we had decided that we were going to link all products to just /product/:id (product show) instead of nesting the links (as per the example we were going off). Definitely willing to open it up for discussion, though! 

(Also realized I was doing single comments instead of starting a review #sorry for the emails)

File: app/views/users/myaccount.html.erb [hunk 4]
```diff
@@ -29,9 +29,9 @@
         <td><%= product.quantity%>
         <td><%= link_to "Edit Product", edit_product_path(product.id)%>
         <% if product.retired == false%>
-          <td><%= link_to "Retire", retire_path(product.id), method: :patch%>
+          <td><%= link_to "Retire", retire_path(product.id), method: 'retire'%>
```
Comment (jessicacaley) [inline app/views/users/myaccount.html.erb]: I believe the method needs to be an HTTP verb as a symbol like :post, :patch, :delete, etc.

File: app/views/users/myaccount.html.erb [hunk 5]
```diff
@@ -29,9 +29,9 @@
...
         <%else%>
-          <td><%= link_to "Reinstate", retire_path(product.id), method: :patch%>
+          <td><%= link_to "Reinstate", retire_path(product.id), method: 'retire'%>
```
Comment (jessicacaley) [inline app/views/users/myaccount.html.erb]: (Same here with the method)

File: test/controllers/products_controller_test.rb [hunk 6]
```diff
@@ -238,21 +241,27 @@
...
 
-      describe "update as Guest" do
-        it "should respond with found and error message to prompt user to login" do
+      describe 'update as Guest' do
+        it 'should respond with found and error message to prompt user to login' do
           patch product_path(Product.first.id)
 
           must_respond_with :found
-          expect(flash[:error]).must_equal "You must be logged in to do this action"
+          expect(flash[:error]).must_equal 'You must be logged in to do this action'
```
Comment (jessicacaley) [inline test/controllers/products_controller_test.rb:249]: Random: was changing all the double quotes to single quotes and {}s to do ends intentional? Both valid; I'm just curious 🤷🏼‍♀️

File: test/controllers/products_controller_test.rb [hunk 7]
```diff
@@ -238,21 +241,27 @@
...
     end
   end
 
-  # it "should get retire" do
-  #   get products_retire_url
-  #   value(response).must_be :success?
-  # end
+  describe "retire" do
+
+    it "should change the retired status of the product" do
```
Comment (jessicacaley) [inline test/controllers/products_controller_test.rb]: Edge cases? (not logged in, non-existant product)

File: config/routes.rb [hunk 8]
```diff
@@ -16,7 +16,8 @@
   resources :products, except: [:destroy] do
     resources :reviews, only: [:new, :create]
   end
-  patch "/products/retire", to: "products#retire", as: "retire"
+
+  post "/products/:id/retire", to: "products#retire", as: "retire"
```
Comment (jessicacaley) [inline config/routes.rb]: I think this route change makes sense! Would love for you to explain the switch from patch to post to me (I trust you that that's right, just want to understand the route better)

**Review by jessicacaley — CHANGES_REQUESTED** [review]
Nice work!! I think the view logic checks out and the toggle functionality is well done. Just have a couple of notes about things like the method verbs and edge case tests -- nothing big, but can definitely talk it over tomorrow if we're not on the same page with those.

File: test/controllers/products_controller_test.rb (hunk 6 above)
Comment (aribray) [reply test/controllers/products_controller_test.rb:249]: I have some extension in VS code that does that, and I can't figure out which one it is/how to turn it off haha

File: config/routes.rb (hunk 8 above)
Comment (aribray) [reply config/routes.rb]: The 'patch' to 'post' was tricky! I had it originally as patch, but it couldn't find the route. When I switched it to post, that worked. I had to do it that way in my tasklist also

File: app/views/categories/show.html.erb (hunk 2 above)
Comment (aribray) [reply app/views/categories/show.html.erb:7]: Good idea! Amy's fixed this, she says

File: app/views/products/index.html.erb (hunk 3 above)
Comment (aribray) [reply app/views/products/index.html.erb]: Hm...I think Mina worked on this method. I think the product_path(product.id) is the correct path. I'm not sure where that comment came from haha

File: app/views/users/myaccount.html.erb (hunk 4 above)
Comment (aribray) [reply app/views/users/myaccount.html.erb]: Idk why this wasn't working before, but I've changed all of the routes/method actions to patch and it still works! I'll commit those changes

File: app/views/users/myaccount.html.erb (hunk 5 above)
Comment (aribray) [reply app/views/users/myaccount.html.erb]: Fixed!

File: config/routes.rb (hunk 8 above)
Comment (aribray) [reply config/routes.rb]: I changed it to patch and it still works! Resolving this comment.

File: test/controllers/products_controller_test.rb (hunk 7 above)
Comment (aribray) [reply test/controllers/products_controller_test.rb]: Got it! I'll add these during project time.

Comment (aribray) [reply test/controllers/products_controller_test.rb]: added not logged in and nonexistent product edge cases!

File: app/views/products/index.html.erb (hunk 3 above)
Comment (jessicacaley) [reply app/views/products/index.html.erb]: Okay, I'll follow up with Mina on this! It looks like we can probably just delete the line but I'll double-check with her.

**Review by jessicacaley — APPROVED** [review]
Looks great!!
